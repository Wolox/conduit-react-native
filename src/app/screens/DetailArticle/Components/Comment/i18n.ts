import i18next from 'i18next';

i18next.addResources('en', 'COMMENTS', {
  VIEW_MORE: 'view more',
  VIEW_LESS: 'view less'
});
