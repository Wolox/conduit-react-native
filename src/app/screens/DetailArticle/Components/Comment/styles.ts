import { StyleSheet } from 'react-native';
import { white } from '@constants/colors';
import { SPACING } from '@constants/dimentions';

const ALIGN_CENTER = 'center';
const COMMON_TWENTY = 20;
const COMMON_TEN = 10;
const COMMON_SIZE_TRASH = 18;

const styles = StyleSheet.create({
  comment: {
    flexDirection: 'row',
    width: '92%',
    alignSelf: ALIGN_CENTER,
    padding: COMMON_TEN,
    backgroundColor: white,
    justifyContent: 'space-between',
    borderRadius: COMMON_TWENTY,
    marginBottom: COMMON_TEN
  },
  containerUser: {
    width: '20%'
  },
  image: {
    height: COMMON_TWENTY,
    width: COMMON_TWENTY,
    borderRadius: COMMON_TWENTY,
    alignSelf: ALIGN_CENTER
  },
  userName: {
    fontSize: COMMON_TEN,
    fontWeight: 'bold',
    width: '100%',
    marginBottom: SPACING.XXXSMALL,
    textAlign: ALIGN_CENTER,
    paddingHorizontal: SPACING.XXSMALL
  },
  commentDate: {
    fontSize: COMMON_TEN,
    textAlign: ALIGN_CENTER
  },
  commentBody: {
    width: '68%'
  },
  commentText: {
    fontSize: 12
  },
  containerImage: {
    borderRadius: COMMON_TWENTY,
    overflow: 'hidden',
    marginBottom: SPACING.XXXSMALL
  },
  containerTrash: {
    justifyContent: 'center',
    alignItems: 'center',
    paddingHorizontal: 5
  },
  icTrash: {
    height: COMMON_SIZE_TRASH,
    width: COMMON_SIZE_TRASH
  }
});

export default styles;
