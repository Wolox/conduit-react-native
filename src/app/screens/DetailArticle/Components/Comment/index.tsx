import React, { useState, useEffect } from 'react';
import { View, Text, Image, TouchableOpacity } from 'react-native';
import i18next from 'i18next';
import { useSelector } from 'react-redux';
import icDefaultArticleImage from '@assets/icons/icDefaultArticleImage.jpg';
import icTrash from '@assets/icons/icTrash.png';
import CustomTextPressable from '@components/CustomTextPressable';
import CustomText from '@components/CustomText';
import { State } from '@interfaces/reduxInterfaces';
import { iComment } from '@interfaces/commentInterfaces';
import { formatDate } from '@utils/dateUtils';

import CustomModal from '../CustomModal';

import styles from './styles';
import './i18n';

interface Props {
  commentContent: iComment;
}

export default function Comment({
  commentContent: {
    author: { username },
    body,
    id,
    createdAt
  }
}: Props) {
  const [presseded, setPresseded] = useState<boolean>(false);
  const [commentText, setCommentText] = useState<string>('');
  const [showModal, setShowModal] = useState<boolean>(false);
  const currentUser = useSelector((state: State) => state?.auth?.currentUser?.username);
  useEffect(() => {
    const handleComment = async () => {
      if (presseded) {
        await setCommentText(body);
      } else {
        await setCommentText(body.substring(0, 78));
      }
    };
    handleComment();
  }, [body, presseded]);
  const onPress = () => setShowModal(!showModal);
  const renderTrash = () => (
    <>
      {username && currentUser === username && (
        <TouchableOpacity style={styles.containerTrash} onPress={onPress}>
          <Image source={icTrash} resizeMode="contain" resizeMethod="scale" style={styles.icTrash} />
        </TouchableOpacity>
      )}
    </>
  );
  return (
    <View style={styles.container}>
      <View style={styles.content}>
        <CustomModal showModal={showModal} setShowModal={setShowModal} idComment={id} />
        <View style={styles.containerUser}>
          <View style={styles.containerImage}>
            <Image style={styles.image} resizeMode="contain" source={icDefaultArticleImage} />
          </View>
          <Text style={styles.userName}>{username}</Text>
          <Text style={styles.commentDate}>{formatDate(createdAt)}</Text>
        </View>
        <View style={styles.commentBody}>
          <CustomText style={[styles.commentText, currentUser !== username && styles.fullCommentBody]}>
            {commentText}
          </CustomText>
          {body.length > 70 && (
            <View>
              <CustomTextPressable
                text={presseded ? i18next.t('COMMENTS:VIEW_LESS') : i18next.t('COMMENTS:VIEW_MORE')}
                onPress={() => setPresseded(!presseded)}
              />
            </View>
          )}
        </View>
        {renderTrash()}
      </View>
    </View>
  );
}
