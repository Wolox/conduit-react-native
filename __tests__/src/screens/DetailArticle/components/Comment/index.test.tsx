import React from 'react';
import renderer from 'react-test-renderer';
import Comment from '@screens/DetailArticle/components/Comment';
import userIcon from '@assets/Profile/icUser.png';

jest.mock('react-redux', () => {
  return {
    useSelector: jest.fn().mockImplementation(() => ({ currentUser: '' })),
    useDispatch: () => jest.fn()
  };
});

describe('Comment', () => {
  test('Comment Snapshot', () => {
    const item = {
      commentContent: {
        author: { username: 'user', image: userIcon },
        body: '',
        id: 1,
        createdAt: ''
      },
      articleSlug: ''
    };
    const tree = renderer.create(<Comment {...item} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
