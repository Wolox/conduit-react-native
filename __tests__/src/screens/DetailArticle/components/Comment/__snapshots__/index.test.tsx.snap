// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment Comment Snapshot 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "backgroundColor": "#FFF",
      "borderRadius": 20,
      "flexDirection": "row",
      "marginBottom": 10,
      "padding": 10,
      "width": "92%",
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "flexDirection": "row",
        "width": "100%",
      }
    }
  >
    <Modal
      animationType="fade"
      hardwareAccelerated={false}
      transparent={true}
      visible={false}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#00000033",
            "flex": 1,
            "justifyContent": "center",
            "marginTop": 22,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#FFF",
              "borderRadius": 20,
              "elevation": 5,
              "height": 180,
              "margin": 20,
              "padding": 35,
              "shadowColor": "#555555",
              "shadowOffset": Object {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.3,
              "shadowRadius": 4,
              "width": "93%",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "color": "#000",
                  "fontFamily": "Nunito",
                  "fontSize": 25.142857142857142,
                  "fontStyle": "normal",
                  "fontWeight": "400",
                },
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "Nunito",
                    "fontSize": 25.142857142857142,
                    "fontStyle": "normal",
                    "fontWeight": "700",
                  },
                ],
                Object {
                  "marginBottom": 15,
                  "textAlign": "center",
                },
              ]
            }
          >
            CUSTOM_COMMENT_MODAL:TITLE
          </Text>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-evenly",
                "marginTop": 20,
                "width": "100%",
              }
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "borderRadius": 20,
                    "elevation": 2,
                    "padding": 15,
                  },
                  Object {
                    "backgroundColor": "#5CB85C",
                  },
                ]
              }
            >
              <Text
                style={
                  Object {
                    "color": "#FFF",
                    "fontFamily": "Nunito",
                    "fontWeight": "bold",
                    "textAlign": "center",
                  }
                }
              >
                CUSTOM_COMMENT_MODAL:CANCEL
              </Text>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "borderRadius": 20,
                    "elevation": 2,
                    "padding": 15,
                  },
                  Object {
                    "backgroundColor": "#FF2400",
                  },
                ]
              }
            >
              <Text
                style={
                  Object {
                    "color": "#FFF",
                    "fontFamily": "Nunito",
                    "fontWeight": "bold",
                    "textAlign": "center",
                  }
                }
              >
                CUSTOM_COMMENT_MODAL:DELETE
              </Text>
            </View>
          </View>
        </View>
      </View>
    </Modal>
    <View
      style={
        Object {
          "width": "24%",
        }
      }
    >
      <View
        style={
          Object {
            "borderRadius": 20,
            "marginBottom": 2,
            "overflow": "hidden",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "process": [Function],
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "borderRadius": 20,
              "height": 20,
              "width": 20,
            }
          }
        />
      </View>
      <Text
        style={
          Object {
            "fontFamily": "Nunito",
            "fontSize": 10,
            "fontWeight": "bold",
            "marginBottom": 2,
            "paddingHorizontal": 4,
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        user
      </Text>
      <Text
        style={
          Object {
            "fontFamily": "Nunito",
            "fontSize": 10,
            "textAlign": "center",
          }
        }
      >
        Invalid Date
      </Text>
    </View>
    <View
      style={
        Object {
          "textAlign": "left",
          "width": "68%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#000",
              "fontFamily": "Nunito",
              "fontSize": 25.142857142857142,
              "fontStyle": "normal",
              "fontWeight": "400",
            },
            Array [],
            Array [
              Object {
                "fontSize": 12,
              },
              Object {
                "width": "100%",
              },
            ],
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;
